src/ssl.c

```diff
@@ -48370,6 +48370,357 @@ int wolfSSL_PEM_write_bio_PKCS7(WOLFSSL_BIO* bio, PKCS7* p7)
+            }
+            else if (section[sectionLen] == '\r') sectionLen--;
+            section[sectionLen+1] = '\0';
+
+            *bcont = wolfSSL_BIO_new(wolfSSL_BIO_s_mem());
+            ret = wolfSSL_BIO_write(*bcont, section, (int)XSTRLEN(section));
+            if (ret != (int)XSTRLEN(section)) return NULL;
+
+            MIME_free_headers(allHeaders);
+            ret = MIME_parse_headers(in, remainLen, &allHeaders);
+            if (ret < 0) {
+                WOLFSSL_MSG("Parsing MIME headers failed.\n");
+                return NULL;
+            }
+            curHeader = MIME_find_header_name("Content-Type", allHeaders);
+            if (!curHeader || (XSTRNCMP(curHeader->body, "application/pkcs7-signature", XSTRLEN("application/pkcs7-signature")) &&
+                               XSTRNCMP(curHeader->body, "application/x-pkcs7-signature", XSTRLEN("application/x-pkcs7-signature")))) {
+                WOLFSSL_MSG("S/MIME headers not found inside multipart message.\n");
+                return NULL;
+            }
+
+            section[0] = '\0';
+            sectionLen = 0;
+            lineLen = wolfSSL_BIO_gets(in, section, remainLen);
+            while(XSTRNCMP(&section[sectionLen], boundary, boundLen) && remainLen > 0) {
+                sectionLen += lineLen;
+                remainLen -= lineLen;
+                lineLen = wolfSSL_BIO_gets(in, &section[sectionLen], remainLen);
+                if (lineLen<=0) return NULL;
+            }
+
+            if (boundary) XFREE(boundary, NULL, DYNAMIC_TYPE_PKCS7);
+        }
+    }
+    else if (curHeader && (!XSTRNCMP(curHeader->body, "application/pkcs7-mime", XSTRLEN("application/pkcs7-mime")) ||
+                           !XSTRNCMP(curHeader->body, "application/x-pkcs7-mime", XSTRLEN("application/x-pkcs7-mime")))) {
+        sectionLen = wolfSSL_BIO_get_len(in);
+        if (sectionLen <= 0) return NULL;
+        section = XMALLOC((sectionLen+1)*sizeof(char), NULL, DYNAMIC_TYPE_PKCS7);
+        if (!section) return NULL;
+        ret = wolfSSL_BIO_read(in, section, sectionLen);
+        if (ret < 0 || ret != sectionLen) {
+            WOLFSSL_MSG("Error reading input BIO.\n");
+            return NULL;
+        }
+    }
+    else {
+        WOLFSSL_MSG("S/MIME headers not found.\n");
+        return NULL;
+    }
+
+    curHeader = MIME_find_header_name("Content-Transfer-Encoding", allHeaders);
+    if (!curHeader) WOLFSSL_MSG("Content-Transfer-Encoding header not found, assuming base64 encoding.");
+    else if (XSTRNCMP(curHeader->body, "base64", XSTRLEN("base64"))) {
+        WOLFSSL_MSG("S/MIME encodings other than base64 not currently supported.\n");
+        return NULL;
+    }
+
+    if (!section || sectionLen <= 0) return NULL;
+    /* Strip trailing newlines. */
+    while ((section[sectionLen-1] == '\r' || section[sectionLen-1] == '\n') && sectionLen > 0) sectionLen--;
+    section[sectionLen] = '\0';
+
+    word32 outLen = ((sectionLen*3+3)/4)+1;
```

Comment:
As per coding standards, please declare all variables at top of function unless only used in block scope.